File: lib/gemini.ts
import { getTemplateById } from '@/components/template-options'
import { loadAiConfig, maskApiKey } from '@/lib/ai-config'

type GeminiPart =
  | { text?: string }
  | {
      inline_data?: {
        mime_type?: string
        data?: string
      }
      inlineData?: {
        mimeType?: string
        data?: string
      }
    }

type GeminiCandidate = {
  content?: {
    parts?: GeminiPart[]
  }
}

type GeminiPayload = {
  candidates?: GeminiCandidate[]
  error?: {
    message?: string
  }
}

export function isTimeoutError(error: unknown) {
  return error instanceof DOMException && error.name === 'TimeoutError'
}

export function buildAspectRatio(templateId: string) {
  const template = getTemplateById(templateId)
  const aspectRatioMap: Record<string, string> = {
    amazon: '1:1',
    detail: '3:4',
    poster: '16:9',
  }

  return aspectRatioMap[template.id] ?? '1:1'
}

export function buildPrompt(templateId: string, prompt?: string) {
  const template = getTemplateById(templateId)
  const basePrompt = prompt?.trim()

  if (!basePrompt) {
    return `生成一张${template.name}风格图片，适合电商展示，画面干净、主体清晰。`
  }

  return `请生成一张${template.name}风格图片。要求：${basePrompt}`
}

function pickImagePart(candidates: GeminiCandidate[] = []) {
  for (const candidate of candidates) {
    const parts = candidate.content?.parts ?? []
    for (const part of parts) {
      const inlineData =
        'inline_data' in part
          ? part.inline_data
          : 'inlineData' in part
            ? part.inlineData
            : undefined

      const normalizedInlineData = inlineData as
        | { mime_type?: string; mimeType?: string; data?: string }
        | undefined

      const mimeType = normalizedInlineData?.mime_type ?? normalizedInlineData?.mimeType
      const data = normalizedInlineData?.data

      if (mimeType?.startsWith('image/') && data) {
        return `data:${mimeType};base64,${data}`
      }
    }
  }

  return ''
}

async function requestGeminiImage(
  modelId: string,
  domain: string,
  apiKey: string,
  prompt: string,
  aspectRatio: string,
  requestId: string,
  attempt: number,
) {
  const startedAt = Date.now()
  console.log(`[gemini][${requestId}] upstream request started`, {
    attempt,
    modelId,
    domain,
    aspectRatio,
  })

  const response = await fetch(`${domain}/v1beta/models/${modelId}:generateContent`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'x-goog-api-key': apiKey,
    },
    body: JSON.stringify({
      contents: [
        {
          parts: [
            {
              text: prompt,
            },
          ],
        },
      ],
      generationConfig: {
        imageConfig: {
          aspectRatio,
        },
      },
    }),
    signal: AbortSignal.timeout(60000),
  })

  const payload = (await response.json()) as GeminiPayload
  console.log(`[gemini][${requestId}] upstream request finished`, {
    attempt,
    status: response.status,
    elapsedMs: Date.now() - startedAt,
    hasCandidates: Boolean(payload.candidates?.length),
  })

  return {
    response,
    payload,
  }
}

async function requestWithRetry(
  modelId: string,
  domain: string,
  apiKey: string,
  prompt: string,
  aspectRatio: string,
  requestId: string,
) {
  try {
    return await requestGeminiImage(modelId, domain, apiKey, prompt, aspectRatio, requestId, 1)
  } catch (error) {
    if (!isTimeoutError(error)) throw error

    console.warn(`[gemini][${requestId}] attempt 1 timed out after 60000ms, retrying once`)

    try {
      return await requestGeminiImage(modelId, domain, apiKey, prompt, aspectRatio, requestId, 2)
    } catch (retryError) {
      if (isTimeoutError(retryError)) {
        console.error(`[gemini][${requestId}] attempt 2 also timed out after 60000ms`)
      }
      throw retryError
    }
  }
}

export async function generateImageForTask(templateId: string, prompt: string, requestId: string) {
  const config = await loadAiConfig()
  const aspectRatio = buildAspectRatio(templateId)
  const finalPrompt = buildPrompt(templateId, prompt)

  const { response, payload } = await requestWithRetry(
    config.modelId,
    config.domain,
    config.apiKey,
    finalPrompt,
    aspectRatio,
    requestId,
  )

  if (!response.ok) {
    throw new Error(payload.error?.message ?? 'Gemini 请求失败，请检查配置文件。')
  }

  const resultUrl = pickImagePart(payload.candidates)

  if (!resultUrl) {
    throw new Error('请求已发出，但模型没有返回图片结果。请检查当前模型是否支持文生图。')
  }

  return {
    resultUrl,
    meta: {
      brand: '图鲸',
      modelId: config.modelId,
      domain: config.domain,
      apiKeyMasked: maskApiKey(config.apiKey),
      requestId,
    },
  }
}

File: backend/app/services/gemini.py
"""Gemini image generation integration used by the FastAPI backend."""

from __future__ import annotations

import logging
import time

import httpx

from app.config import load_ai_config, mask_api_key
from app.services.template import build_aspect_ratio, build_prompt

logger = logging.getLogger(__name__)
GEMINI_TIMEOUT = httpx.Timeout(connect=10.0, read=180.0, write=30.0, pool=30.0)


def _pick_image_part(candidates: list[dict]) -> str:
    for candidate in candidates:
        parts = candidate.get("content", {}).get("parts", [])
        for part in parts:
            inline = part.get("inline_data") or part.get("inlineData") or {}
            mime = inline.get("mime_type") or inline.get("mimeType", "")
            data = inline.get("data", "")
            if mime.startswith("image/") and data:
                return f"data:{mime};base64,{data}"
    return ""


async def _request_gemini(
    client: httpx.AsyncClient,
    model_id: str,
    domain: str,
    api_key: str,
    prompt: str,
    aspect_ratio: str,
    request_id: str,
    attempt: int,
) -> dict:
    started = time.monotonic()
    logger.info("[gemini][%s] request started attempt=%d model=%s", request_id, attempt, model_id)

    url = f"{domain}/v1beta/models/{model_id}:generateContent"
    resp = await client.post(
        url,
        headers={"Content-Type": "application/json", "x-goog-api-key": api_key},
        json={
            "contents": [{"parts": [{"text": prompt}]}],
            "generationConfig": {"imageConfig": {"aspectRatio": aspect_ratio}},
        },
    )

    try:
        payload = resp.json()
    except ValueError:
        payload = {"error": {"message": "上游服务返回了无法解析的响应。"}}
    elapsed = int((time.monotonic() - started) * 1000)
    logger.info("[gemini][%s] finished attempt=%d status=%d elapsed=%dms", request_id, attempt, resp.status_code, elapsed)
    return {"status_code": resp.status_code, "payload": payload}


async def _request_with_timeout(
    model_id: str, domain: str, api_key: str, prompt: str, aspect_ratio: str, request_id: str,
) -> dict:
    async with httpx.AsyncClient(timeout=GEMINI_TIMEOUT) as client:
        return await _request_gemini(client, model_id, domain, api_key, prompt, aspect_ratio, request_id, 1)


async def generate_image_for_task(
    template_id: str,
    prompt: str,
    request_id: str,
    model_id: str | None = None,
) -> dict:
    config = load_ai_config()
    model = config.get_model(model_id or config.default_model)
    aspect_ratio = build_aspect_ratio(template_id)
    final_prompt = build_prompt(template_id, prompt)

    try:
        result = await _request_with_timeout(
            model.id, model.domain, model.api_key, final_prompt, aspect_ratio, request_id,
        )
    except httpx.TimeoutException as exc:
        logger.warning("[gemini][%s] upstream timeout", request_id)
        raise RuntimeError(
            "请求已发出，但等待上游模型响应超时。当前任务不会继续等待该结果，请稍后重试。"
        ) from exc
    except httpx.RequestError as exc:
        logger.warning("[gemini][%s] request failed: %s", request_id, exc)
        raise RuntimeError("请求发送失败，当前无法连接上游模型服务，请稍后重试。") from exc

    if result["status_code"] >= 400:
        error_msg = result["payload"].get("error", {}).get("message", "Gemini 请求失败，请检查配置文件。")
        raise RuntimeError(error_msg)

    result_url = _pick_image_part(result["payload"].get("candidates", []))
    if not result_url:
        raise RuntimeError("请求已发出，但模型没有返回图片结果。请检查当前模型是否支持文生图。")

    return {
        "resultUrl": result_url,
        "meta": {
            "brand": "图鲸",
            "modelId": model.id,
            "domain": model.domain,
            "apiKeyMasked": mask_api_key(model.api_key),
            "requestId": request_id,
        },
    }
